miniapps/con_gen/connection_generator.hpp

```diff
@@ -0,0 +1,261 @@
+        std::vector<projection> const connectome)
+        : connectome_(connectome) {
+        cell_gid_type gid_idx = 0;
+
+        // Create the local populations with start index set
+        for (auto pop : populations) {
+            populations_.push_back(population_indexed(
+                pop.x_side, pop.y_side, pop.periodic, gid_idx ));
+
+            gid_idx += pop.n_cells;
+        }
+    }
+
+    // Returns a vector of all synaptic parameters sets for this gid
+    std::vector<synaps_pars> synapses_on(cell_gid_type gid) {
+        std::mt19937 gen;
+        gen.seed(gid);
+
+        std::vector<synaps_pars> connections;
+        for (auto project : connectome_)
+        {
+            // Sanity check that the populations exist
+            EXPECTS(project.pre_idx < populations_.size());
+            EXPECTS(project.post_idx < populations_.size());
+
+            // Shorthand for the pre and post populations
+            auto pre_pop = populations_[project.pre_idx];
+            auto post_pop = populations_[project.post_idx];
+            auto pro_pars = project.pars;
+
+            // Distribution to draw the weights
+            std::normal_distribution<float> weight_distr(pro_pars.weight_mean, pro_pars.weight_sd);
+
+            // Used to assure correct weight sign
+            float weight_sign = pro_pars.weight_mean < 0 ? -1 : 1;
+
+            // Check if this gid receives connections via this projection
+            // TODO: Replace with the fancy in range function we have somewhere in the utils
+            if (gid < post_pop.start_index || gid > (post_pop.start_index + post_pop.n_cells)) {
+                continue;
+            }
+
+            // Convert to the local gid of the post neuron
+            auto pop_local_gid = gid - post_pop.start_index;
+
+            // Convert this to a normalized location
+            std::pair<float, float> post_location = {
+                float(pop_local_gid % post_pop.x_side) / post_pop.x_side,
+                float(pop_local_gid / post_pop.y_side) / post_pop.y_side};
+
+            // If we have non square sides we need to correct the stdev to get
+            // circular projections!
+            float sd_x = pro_pars.sd;
+            float sd_y = pro_pars.sd;
+            if (post_pop.x_side != post_pop.y_side) {
+                if (post_pop.x_side < post_pop.y_side) {
+                    float ratio = float(post_pop.y_side) / post_pop.x_side;
+                    sd_x *= ratio;
+                }
+                else {
+                    float ratio = float(post_pop.x_side) / post_pop.y_side;
+                    sd_y *= ratio;
+                }
+            }
+
+            // Now we sample from the pre population based on the x,y location
+            // We supply the connection_impl with the size of the pre population
+            auto pre_locations = connection_locations(gen, post_location,
+                pro_pars.count, sd_x, sd_y, post_pop.periodic);
+
```

Comment:
I would define a little 2D point type to use instead of a `std::pairi<float,float>`, to clean up the code:

```
struct point {
    float x;
    float y;
};

float distance(const point& p1, const point& p2) {
    return std::sqrt(arb::square(p1.x-p2.x)+arb::square(p1.y-p2.y));                                     
}
```

Reply:
I had that also in mind. I did it now